pageserver/src/tenant/timeline.rs

```diff
@@ -873,29 +873,59 @@ impl Timeline {
     pub async fn evict_layer(&self, layer_file_name: &str) -> anyhow::Result<Option<bool>> {
         let Some(local_layer) = self.find_layer(layer_file_name) else { return Ok(None) };
         let Some(remote_client) = &self.remote_client else { return Ok(Some(false)) };
-        self.evict_layer_impl(local_layer, remote_client)
-            .await
-            .map(|res| Some(res))
+        let results = self
+            .evict_layer_batch(remote_client, &[local_layer])
+            .await?;
+        assert_eq!(results.len(), 1);
+        let result: Option<anyhow::Result<bool>> = results.into_iter().next().unwrap();
+        match result {
+            None => anyhow::bail!("task_mgr shutdown requested"),
+            Some(Ok(b)) => Ok(Some(b)),
+            Some(Err(e)) => Err(e),
+        }
     }
 
     // The `remote_client` should be this timeline's `self.remote_client`.
     // We make the caller provide it so that they are responsible for handling the case
     // where someone wants to evict the layer but no remote storage is configured.
-    async fn evict_layer_impl(
+    async fn evict_layer_batch(
         &self,
-        local_layer: Arc<dyn PersistentLayer>,
         remote_client: &Arc<RemoteTimelineClient>,
-    ) -> anyhow::Result<bool> {
-        if local_layer.is_remote_layer() {
-            return Ok(false);
-        }
-
-        // ensure the current layer is uploaded for sure
+        layers: &[Arc<dyn PersistentLayer>],
+    ) -> anyhow::Result<Vec<Option<anyhow::Result<bool>>>> {
+        // ensure that the layers have finished uploading
+        // (don't hold the layer_removal_cs while we do it, we're not removing anything yet)
         remote_client
             .wait_completion()
             .await
             .context("wait for layer upload ops to complete")?;
 
+        // now lock out layer removal (compaction, gc, timeline deletion)
+        let layer_removal_guard = self.layer_removal_cs.lock().await;
+
+        let mut results = Vec::with_capacity(layers.len());
+
+        for l in layers.iter() {
+            let res = if task_mgr::is_shutdown_requested() {
+                None
+            } else {
+                Some(self.evict_layer_batch_impl(&layer_removal_guard, l).await)
+            };
+            results.push(res);
+        }
+
+        Ok(results)
+    }
+
+    async fn evict_layer_batch_impl(
```

Comment:
This is no longer async right?

IMHO the `layers` could be kept locked as well, since the other lock is already. Also, only one batch should happen. Or perhaps there's a reason why multiple batches?

Reply:
> This is no longer async right?

Right. Will remove it for now and keep using the synchronous unlink code.

> IMHO the layers could be kept locked as well, since the other lock is already.Also, only one batch should happen. Or perhaps there's a reason why multiple batches?

Yeah. Both were oversights from my side, sorry about that.